webapp/src/App/Pages/Settings/Tokens/TokenModal.tsx

```diff
@@ -0,0 +1,188 @@
+            return ;
+        //2) Create a new role with type == READER
+        const newRole : UserRoleDTO = {
+            id: selectedGroup.id,
+            name: selectedGroup.name,
+            role: RoleType.READER // READER by default
+        }
+        //3) Add the role in roleList
+        setRoleList(roleList.concat([newRole]));
+    }
+
+    const deleteRoleFromList = (group_id: string) => {
+        // Delete role from roleList
+        const tmpList = roleList.filter((item) => item.id !== group_id);
+        setRoleList(tmpList);
+    }
+
+    // Update Role in roleList
+    const updateRoleFromList = (group_id: string, role : RoleType) => {
+        // 1) Find the role
+        const tmpList : UserRoleDTO[] = ([] as UserRoleDTO[]).concat(roleList);
+        const index = roleList.findIndex((item) => item.id === group_id);
+        if(index >= 0)
+        {
+            // 2) Update role in roleList
+            tmpList[index].role = role;
+            setRoleList(tmpList);
+        }
+    }
+
+    const onSave = async () => {
+        try{
+            // 1) Create new user or delete previous role list in DB
+            let tmpTokenId = await prepareDBForRole(tokenName,
+                                                    checked,
+                                                    initialRoleList,
+                                                    onNewTokenCreation,
+                                                    onUpdateToken,
+                                                    tokenInfos);
+            // 2) Create roles
+            await createNewRoles(tmpTokenId, roleList);
+            setInitialRoleList(roleList);
+            enqueueSnackbar(t('settings:onTokenUpdate'), { variant: 'success' });
+        }
+        catch(e)
+        {
+            enqueueSnackbar(t('settings:onTokenSaveError'), { variant: 'error' });
+        }
+    }
+
+    useEffect(() => {
+        const init = async () =>{
+            try {
+              
+                // Get list of all groups and add it to groupList or locally from access_token
+                const groups = await getGroups();
+                const filteredGroup = groups.filter((item) => item.id !== "admin");
+                
+                setGroupList(filteredGroup);
+                if(filteredGroup.length > 0)
+                    setActiveGroup(filteredGroup[0]);
+                // 2) If tokenInfos exist => get list of user roles and update roleList
+                if(!!tokenInfos)
+                {
+                    const tokens = await getBotsInfos(tokenInfos.id);
+                    const filteredRoles = tokens.roles.filter((item) => item.id !== "admin")
+                    console.log(filteredRoles)
+                    setRoleList(filteredRoles);
+                    setInitialRoleList(filteredRoles);
+                    setTokenName(tokenInfos.name);
+                    setChecked(tokenInfos.isAuthor);
```

Comment:
here the modal should be in readonly (see previous comment about preventing update)